hera_cal/abscal.py

```diff
@@ -735,39 +787,60 @@ def global_phase_slope_logcal(model, data, antpos, wgts=None, refant=None, verbo
     avg_model, _, _ = avg_data_across_red_bls(DataContainer({k: model[k] for k in keys}),
                                               antpos, wgts=wgts, broadcast_wgts=False, tol=tol, reds=reds)
 
-    # build linear system
-    ls_data, ls_wgts = {}, {}
+    ls_data, ls_wgts, bls, pols = {}, {}, {}, {}
     for rk in red_keys:
         # build equation string
         eqn_str = '{}*Phi_ew_{} + {}*Phi_ns_{} - {}*Phi_ew_{} - {}*Phi_ns_{}'
         eqn_str = eqn_str.format(antpos[rk[0]][0], split_pol(rk[2])[0], antpos[rk[0]][1], split_pol(rk[2])[0],
                                  antpos[rk[1]][0], split_pol(rk[2])[1], antpos[rk[1]][1], split_pol(rk[2])[1])
+        bls[eqn_str] = antpos[rk[0]] - antpos[rk[1]]
+        pols[eqn_str] = rk[2]
 
         # calculate median of unflagged angle(data/model)
         # ls_weights are sum of non-binary weights
-        delta_phi = np.angle(avg_data[rk] / avg_model[rk])
+        dm_ratio = avg_data[rk] / avg_model[rk]
+        dm_ratio /= np.abs(dm_ratio)  # This gives all channels roughly equal weight, moderating the effect of RFI (as in firstcal)
         binary_flgs = np.isclose(avg_wgts[rk], 0.0)
-        delta_phi[binary_flgs] *= np.nan
-        avg_wgts[rk][np.isinf(delta_phi) + np.isnan(delta_phi)] = 0.0
-        delta_phi[np.isinf(delta_phi) + np.isnan(delta_phi)] *= np.nan
-        ls_data[eqn_str] = np.nanmedian(delta_phi[:, edge_cut:(delta_phi.shape[1] - edge_cut)], axis=1, keepdims=True)
-        ls_wgts[eqn_str] = np.sum(avg_wgts[rk][:, edge_cut:(delta_phi.shape[1] - edge_cut)], axis=1, keepdims=True)
+        dm_ratio[binary_flgs] *= np.nan
+        avg_wgts[rk][np.isinf(dm_ratio) + np.isnan(dm_ratio)] = 0.0
+        dm_ratio[np.isinf(dm_ratio) + np.isnan(dm_ratio)] *= np.nan
+        if solver == 'linfit':  # we want to fit the angles
+            ls_data[eqn_str] = np.nanmedian(np.angle(dm_ratio[:, edge_cut:(dm_ratio.shape[1] - edge_cut)]), axis=1, keepdims=True)
+        elif solver == 'dft':  # we want the full complex number
+            ls_data[eqn_str] = np.nanmedian(dm_ratio[:, edge_cut:(dm_ratio.shape[1] - edge_cut)], axis=1, keepdims=True)
+        ls_wgts[eqn_str] = np.sum(avg_wgts[rk][:, edge_cut:(dm_ratio.shape[1] - edge_cut)], axis=1, keepdims=True)
 
         # set unobserved data to 0 with 0 weight
         ls_wgts[eqn_str][np.isnan(ls_data[eqn_str])] = 0
         ls_data[eqn_str][np.isnan(ls_data[eqn_str])] = 0
 
-    # setup linsolve and run
-    solver = linsolve.LinearSolver(ls_data, wgts=ls_wgts)
-    echo("...running linsolve", verbose=verbose)
-    fit = solver.solve()
-    echo("...finished linsolve", verbose=verbose)
+    if solver == 'linfit':  # build linear system for phase slopes and solve with linsolve
+        # setup linsolve and run
+        solver = linsolve.LinearSolver(ls_data, wgts=ls_wgts)
+        echo("...running linsolve", verbose=verbose)
+        fit = solver.solve()
+        echo("...finished linsolve", verbose=verbose)
```

Comment:
All of these "if linfit, elif dft" switches are making me wonder if these shouldn't just be two separate subclasses, one that does linfit, one that does dft

Reply:
It's a toss up, but a fair bit of the calculation gets used multiple times. This also makes it easier to use the Abscal.global_phase_slope_logcal and Abscal.custom_phs_slope_gain, since it enforces a universal interface.